### PR TITLE
OBJLoader: Support for faces that has more than 4 vertices

### PR DESCRIPTION
**Pull request by zerohun**
I know there were several discussions about this in #10438, #10917
I understand the concerns about performance in the conversations in #10917
I tried not to damage performance for common use cases (faces with 3 or 4 vertices)

### CODE REVIEW

**Comment by mrdoob** [comment]
Do you mind converting spaces indentations to tabs? 😇

**Comment by zerohun** [comment]
@mrdoob  I just converted it to tabs :)

**Comment by zerohun** [comment]
performance comparison: https://codepen.io/zerohoon/pen/wqzyea

**Comment by zerohun** [comment]
Just made a really small change to improve readability.

File: examples/js/loaders/OBJLoader.js [hunk 1]
```diff
@@ -25,6 +25,13 @@ THREE.OBJLoader = ( function () {
 	// usemtl material_name
 	var material_use_pattern     = /^usemtl /;
 
+	var adding_face_modes = {
+		vertex: 1,
+		vertex_uv: 2,
+		vertex_uv_normal: 3,
```
Comment (WestLangley) [inline examples/js/loaders/OBJLoader.js]: How about
```
var MODE = { ADDING_VERTEX: 1, ADDING_VERTEX_UV: 2, ADDING_VERTEX_UV_NORMAL: 3, ADDING_VERTEX_NORMAL: 4 };
```

File: examples/js/loaders/OBJLoader.js [hunk 2]
```diff
@@ -260,72 +267,96 @@ THREE.OBJLoader = ( function () {
...
 
-				var ia = this.parseVertexIndex( a, vLen );
-				var ib = this.parseVertexIndex( b, vLen );
-				var ic = this.parseVertexIndex( c, vLen );
-				var id;
+				var addingFaceMode = args.pop();
+				var ivs = [ ], ius = [ ], ins = [ ];
 
-				if ( d === undefined ) {
+				if( addingFaceMode === adding_face_modes.vertex ) {
```
Comment (WestLangley) [inline examples/js/loaders/OBJLoader.js]: A switch statement would be more appropriate.
```
var mode = args.pop();

switch ( mode ) {

	case MODE.ADDING_VERTEX:

		. . .

		break;

	case MODE.ADDING_VERTEX_UV:

		. . .

		break;

}

```

Comment (WestLangley) [inline examples/js/loaders/OBJLoader.js]: Also, you can use http://zz85.github.io/mrdoobapproves/ to check your formatting.

**Comment by zerohun** [comment]
@WestLangley  Thank you for your great feedbacks. I revised it as you suggested. :)

File: examples/js/loaders/OBJLoader.js [hunk 3]
```diff
@@ -260,72 +267,103 @@ THREE.OBJLoader = ( function () {
...
-					var nLen = this.normals.length;
-					ia = this.parseNormalIndex( na, nLen );
+						break;
 
-					ib = na === nb ? ia : this.parseNormalIndex( nb, nLen );
-					ic = na === nc ? ia : this.parseNormalIndex( nc, nLen );
+					case MODE.ADDING_VERTEX_UV_NORMAL:
 
-					if ( d === undefined ) {
+						for ( var i = 0; i < args.length - 2; i = i + 3 ) {
```
Comment (WestLangley) [inline examples/js/loaders/OBJLoader.js]:
```
i += 3
```
Fix this pattern everywhere, please.

File: examples/js/loaders/OBJLoader.js [hunk 4]
```diff
@@ -260,72 +267,103 @@ THREE.OBJLoader = ( function () {
...
+							ivs.push( this.parseVertexIndex( args[ i ], vLen ) );
+							ins.push( this.parseNormalIndex( args[ i + 1 ], nLen ) );
+
+						}
+
+						break;
+
+				}
+
+				for ( var j = 2; j < ivs.length; j++ ) {
```
Comment (WestLangley) [inline examples/js/loaders/OBJLoader.js]:
```
j ++
```
Please use http://zz85.github.io/mrdoobapproves/ to check your formatting so someone else doesn't have to do it.

**Comment by WestLangley** [comment]
I have had a look at the code for readability, but someone else will have to evaluate it for performance.

**Comment by mrdoob** [comment]
Do you mind explaining how you're triangulating the ngons?

**Comment by mrdoob** [comment]
Ended up merging #11871, which was similar to this PR but a bit simpler.

**Comment by mrdoob** [comment]
Many thanks for the PR though!

**Comment by zerohun** [comment]
@mrdoob If there is no big difference in performance, I also think simpler solution is better :)

**Pull request closed without merging**